handler.go

```diff
@@ -226,12 +226,9 @@ func makeClientStream(ctx context.Context, stream *yarpctransport.ClientStream,
 }
 
 // makeBidiStream starts bi-directional streaming rpc
-func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream, streamIO StreamIO, opts StreamRequestOptions) error {
+func makeBidiStream(ctx context.Context, cancel context.CancelFunc, stream *yarpctransport.ClientStream, streamIO StreamIO, opts StreamRequestOptions) error {
```

Comment:
(off this commit) Thoughts; do we need to play with cancel in this function? It is very unfamiliar to control execution of grouting through cancel context calls.

I believe that cancel should only be controlled by the caller of this method (as we have for makeServerStream and makeClientStream actually). 

Here, it looks like you are using cancel to stop a for loop from a goroutine. Can we use a channel for this instead?
Maybe something like this:

```
// Start go routine to concurrently send stream messages.
sendErr := make(chan error)
receiveErr := make(chan error)
done := make(chan bool)

go func() {
	var err error
	for err == nil {
		var reqBody []byte
		reqBody, err = streamIO.NextRequest()
		if err == io.EOF {
			err = closeSendStream(ctx, stream)
			break
		}
		if err != nil {
			break
		}
		reqWaiter.wait(ctx)
		err = sendStreamMessage(ctx, stream, reqBody)
	}
	if err != nil {
		sendErr <- err
		close(sendErr)
	}
	close(done)
}()

go func () {
	var err error
	for err == nil {
		var resBody []byte
		resBody, err = receiveStreamMessage(ctx, stream)
		if err != nil {
			break
		}
		err = streamIO.HandleResponse(resBody)
	}
	
	receiveErr <- err
	close(receiveErr)
}()

select {
case err := <-receiveErr:
	if err != io.EOF {
		return err
	}
case err := <-sendErr:
	if err != io.EOF {
		return err
	}
case <- done:
	return nil
}
```
Here, if one fo the goroutine have an error, `makeBidiStream` will then return the error. If sending messages is over,  `makeBidiStream` return nil.

Let me know what you think, I might miss some context here as well.

Reply:
Send and Receive stream APIs both use the same underlying connection/(http-2 stream) created during the stream request initiation. Both send and receive are blocking calls which return when: 
* data has been sent or received successfully 
* error (could be stream closure as well)

The idea in this PR is to close the stream by cancelling the context sent during request initiation, which causes send/receive API call to return immediately as the underlying stream ends on context cancel.

>do we need to play with cancel in this function? It is very unfamiliar to control execution of grouting through cancel context calls.

When blocking operations in two different routines are dependent on an underlying object which is context aware, then cancelling context is the cleanest way of unblocking them. 

> If sending messages is over, makeBidiStream return nil.

Good idea, but I think this might be a bit complicated as we have to wait for both routines to gracefully exit.
We cannot return once sending is over,  number of messages to be sent and received can vary so it's not safe to assume stream is over when the all the messages have been sent. Calling `close(done)` in sender routine would cause early exit of handler and it will cause some messages to be lost from the server as caller of this handler cancels the context once returned.  Another potential issue is routine leak when either of the routine exit early as handler does not wait for both to complete. (there might be deadlock issues when one of the routine finishes it causes select to return err/nil while the other routine also will try to send on a channel which does not have a consumer).

Reply:
Agree, there are some flaws in my proposal. But nothing insurmountable. This is more an off topic of this diff (since it is a refactoring and not a bug fix). Let's sync offline on how we want to change this piece of logic (or if we finally want).